### ISSUE TITLE
ElasticSearch @searchable doesn't index AWSJSON correctly

### ISSUE THREAD

**Issue by troygoode** [aws-amplify/amplify-cli#673]
If a model contains an [`AWSJSON`](https://docs.aws.amazon.com/appsync/latest/devguide/scalars.html) field, adding/updating records using that model will NOT result in those changes being sent to ElasticSearch UNLESS you double-stringify the field. Instead the change is properly recorded in Dynamo and silently (?) fails when the update is being sent to ElasticSearch – the ElasticSearch index just never updates.

In other words, this will not work:

```javascript
const myJSONfield = JSON.stringify({ example: "json" })

await API.graphql(graphqlOperation(mutations.createTodo, {input: {
  name: 'Todo 1',
  description: 'Learn AWS AppSync',,
  myJSONfield: myJSONfield
}))
```

This will:

```javascript
let myJSONfield = JSON.stringify({ example: "json" })
myJSONfield = JSON.stringify(myJSONfield)

await API.graphql(graphqlOperation(mutations.createTodo, {input: {
  name: 'Todo 1',
  description: 'Learn AWS AppSync',,
  myJSONfield: myJSONfield
}))
```

**Comment by UnleashedMind** [comment]
Thanks for the feedbacks. 
The [doc](https://docs.aws.amazon.com/appsync/latest/devguide/scalars.html) specifies that: 
`Invalid JSON strings like {a: 1}, {'a': 1} and Unquoted string will throw GraphQL validation errors.`

In your example, single stringify result in: 
`{"example":"json"}`
and double stringify returns: 
`"{\"example\":\"json\"}"`

So you are right, only double stringify produces the correct format, as specified in the doc.

What can be improved would be to surface this validation error earlier in the workflow. I will mark this as feature request for now and have a discussion about it with the team.

**Comment by troygoode** [comment]
@UnleashedMind Sorry, that isn't quite correct. `Invalid JSON strings` is referring to the fact that `{a: 1}` isn't properly quoted (it should be `{"a": 1}`). In this case the problem is NOT at the AppSync->DynamoDB layer, as the values are properly stored in the DynamoDB table as AWSJSON values and are parsed correctly by Dynamo (as seen when looking at the tuple from the Dynamo UI in the AWS Console). The problem only occurs in the Dynamo->ElasticSearch sync.

In other words, this will store in DynamoDB but cause the entire record to note sync to ElasticSearch:

`"{\"example\":\"json\"}"`

You need to `JSON.stringify` that one more time for it to work correctly.

**Comment by UnleashedMind** [comment]
@troygoode 
I updated my last comment after your last comment, just to put the full paragraph there because at the beginning of that paragraph, there was the sample that put the escape character `\` in front of the quote.
I've asked our JS team to check on this.

**Comment by mikeparisstuff** [comment]
Thanks @troygoode for the feedback. It sounds like the issue may lie in the function that streams information from DynamoDB to Lambda. I've recorded this and will take a look when I can. If you have any relevant logs from the streaming function that may help please share them. I'll make sure that a test case gets added to cover this use case as well.

**Comment by mikeparisstuff** [comment]
Checking in again. I just went through and tried this myself but was unable to reproduce the issue.

I created an API with this schema:

```graphql
type Post @model @searchable {
        id: ID!
        author: String!
        title: String
        content: String
        url: String
        ups: Int
        downs: Int
        version: Int
        relatedPosts: [Post]
        postedAt: String
        comments: [String!]
        ratings: [Int!]
        percentageUp: Float
        isPublished: Boolean
        jsonField: AWSJSON
    }
```

then ran this mutation:

```graphql
mutation M {
  createPost(input: {
    author:"mp",
    title:"a title",
    content: "testing content",
    url: "www.amazon.com",
    ups: 10,
    downs: 5,
    version: 1,
    jsonField: "{\"a\": 1}"
  }) {
    id
    title
    content
    url
    ups
    downs
    version
    jsonField
  }
}
```

which yielded this result:

```json
{
  "data": {
    "createPost": {
      "id": "3f58cf74-ffc1-4efc-ac5d-f42ad9cf4ae2",
      "title": "a title",
      "content": "testing content",
      "url": "www.amazon.com",
      "ups": 10,
      "downs": 5,
      "version": 1,
      "jsonField": "{\"a\":1}"
    }
  }
}
```

I then went to kibana and ran this query:

```
GET _search
{
  "query": {
    "match": {
      "jsonField.a": 1
    }
  }
}
```

and received the correct information as expected. Could you please provide more info and steps that I can use to reproduce this issue?

Thanks

**Comment by elorzafe** [comment]
@troygoode 

I have this workaround, @mikeparisstuff you can correct this in case I am wrong or is a different way to do that.

I modified my schema for searchTodos, my filter is a `AWSJSON` instead of `SearchableTodoFilterInput` like this:

```graphql
type Query {
	getTodo(id: ID!): Todo
	listTodos(filter: ModelTodoFilterInput, limit: Int, nextToken: String): ModelTodoConnection
	searchTodos(
		filter: AWSJSON,
		sort: SearchableTodoSortInput,
		limit: Int,
		nextToken: Int
	): SearchableTodoConnection
}
```

Then I updated my `searchTodos` request mapping template by replacing this line `$util.transform.toElasticsearchQueryDSL($ctx.args.filter)` with this `$util.toJson($ctx.args.filter)`

Finally on my app code I sent the ElasticSearch query as AWSJSON like this

```javascript
    const filter = JSON.stringify({
        "match": {
          "jsonField.a": 1
        }
    });
    const response = await API.graphql({
      query: searchTodos, variables: {
        filter
      }
    });
```

**Comment by kaustavghosh06** [comment]
@troygoode Did @elorzafe answer your question?

**Comment by sakhmedbayev** [comment]
I have the same issue. I have more than 1500 records that I imported to DynamoDB (I did not use graphql for that matter). All the records that do not have `AWSJSON` type field were successfully indexed by ElasticSearch.

**Comment by sakhmedbayev** [comment]
I fixed my issue. Elasticsearch could not properly index my records because records that I imported "manually" to DynamoDB had various data types, e.g. JSON and String. 

Basically one needs to watch that a field has consistent data types across all records.

**Comment by kaustavghosh06** [comment]
Closing this issue as it seems to be solved. Please let me know is this problem still exists.

**Comment by peterv1988** [comment]
> 
> 
> @troygoode
> 
> I have this workaround, @mikeparisstuff you can correct this in case I am wrong or is a different way to do that.
> 
> I modified my schema for searchTodos, my filter is a `AWSJSON` instead of `SearchableTodoFilterInput` like this:
> 
> ```graphql
> type Query {
> 	getTodo(id: ID!): Todo
> 	listTodos(filter: ModelTodoFilterInput, limit: Int, nextToken: String): ModelTodoConnection
> 	searchTodos(
> 		filter: AWSJSON,
> 		sort: SearchableTodoSortInput,
> 		limit: Int,
> 		nextToken: Int
> 	): SearchableTodoConnection
> }
> ```
> 
> Then I updated my `searchTodos` request mapping template by replacing this line `$util.transform.toElasticsearchQueryDSL($ctx.args.filter)` with this `$util.toJson($ctx.args.filter)`
> 
> Finally on my app code I sent the ElasticSearch query as AWSJSON like this
> 
> ```js
>     const filter = JSON.stringify({
>         "match": {
>           "jsonField.a": 1
>         }
>     });
>     const response = await API.graphql({
>       query: searchTodos, variables: {
>         filter
>       }
>     });
> ```

How where you able to modify the existing `searchTodos` query in the `schema.graphql` file? When I to put the *searchTodos* query, with *filter: AWSJSON* exactly as you posted I get following error:
```$ amplify api gql-compile
Error: Object type extension 'Query' cannot redeclare field searchTodos```

I have now manually updated the schema in AppSync, and updated the resolver as you mentioned which I all got working now. But I can't managed to 'fix' this inside the code of my Amplify project as you mentioned.

**Comment by davidbiller** [comment]
@mikeparisstuff 
You did a great job showing that. 

But what about relatedPosts: [Post]?

If I stream an ElasticSearch from my schema file created with @searchable, does it just stream the POST correctly? 

Means in ElasticSearch I have no access to related fields. 

I want to build something like this:
User can swip cards like at tinder. I have to remember the swiped cards somehow. And I do that in an extra table? 
In ElasticSearch I would like to do something like that:
give me all users I have not seen yet, within a radius of 30 km.

Whereby the users then just 
seenFrom: [User]
I would have gotten. Somehow I have to show ElasticSearch which cards the user has already seen. What is the best way to do this?
